Faster episode booting

SolaceEllery opened the pull request:
This is another pull request by me. I decided to make a new way of booting episodes on boot, so rather than use the old method of "booting intro.lvl -> select episode from list -> boot it -> load episode", I decided on bypassing intro.lvl ENTIRELY so now that it's "skip intro.lvl -> load episode".

Check out this video for a loading comparison for Super Mario Bros. 2024, an episode by vitzie629: https://youtu.be/IDYBZzBwuzI

Note: When loading LunaLoader.exe by itself with this PR, it tells you to use the SMBX2 launcher instead to load LunaLoader.exe. This means the final way to use the old legacy launcher is finally bypassed, and any way to use intro.lvl is no longer possible.

Commit:
Wip redo

Commit:
Merge branch 'WohlSoft:master' into faster-episode-booting

Commit:
Merge branch 'WohlSoft:master' into faster-episode-booting

Commit:
Finally works, geez

Commit:
Fix Misc.loadEpisode()

Commit:
Oops

Commit:
Oops again

Commit:
Whoops, again

Commit:
Fix this check

Commit:
Switch isBootingSinglePlayer bools, in case. Also add episodeLoadedOnboot, to make sure that there's a check before doing things for Misc.loadEpisode

Commit:
Add a check for the world path, if not found it doesn't load the episode

Commit:
Oops

Commit:
Oops x2

Commit:
Oops x3

Commit:
Let's shorten and fix this

Commit:
Even MORE checks ugh

Commit:
Meh

Commit:
Idk how to make Misc.loadEpisode work at this point...

Emral commented on LunaDll/Misc/RuntimeHookComponents/RuntimeHookHooks.cpp:
```diff
@@ -1602,6 +1602,27 @@ Characters playerStoredCharacters[] = {CHARACTER_MARIO,CHARACTER_MARIO,CHARACTER
+    std::string nameST = WStr2Str(nameWS);
+
+    //If the legacy title screen is about to boot, prevent that and go straight to loading the episode
+    if(nameST == "intro.lvl" && dirST == gAppPathUTF8)
```
I'm not too familiar with the underlying code of SMBX but shouldn't there be a flag, level mode or specific point of code execution that indicates if we're loading the title screen? Something more robust than going by filename. How does this if check handle exiting an episode via the pause menu or Misc.exitGame()?

SpencerEverly replied:
I don't think there is, unless I'm wrong. This is the earliest we can detect loading the title screen without directly loading intro.lvl.

Also the directory check is there to prevent any other lvl files from other places (Like a world folder) having the same name, be prevented to be loaded.

Supermario1313 replied:
there's `GM_LEVEL_MODE` which is equal to `-1` during the title screen.

Emral commented on LunaDll/GameConfig/GameAutostart.h:
```diff
@@ -29,6 +20,14 @@ class GameAutostart
 
     bool applyAutostart();
+    
+    //Settings
```
Do these really need to be public?

SpencerEverly replied:
The reason why they are public is because it's used for the LaunchEpisode function when booting SMBX2. I could move the LaunchEpisode function though to GameAutostart, if that'll be better

Commit:
Fix autostart level loading (Kinda), and add episode list generation for LunaLua (World only)

Commit:
Ugh

Commit:
Add gLunaLua.exitContext() when loading episode from Lua

Commit:
Remove unused P2 functions, use playerStoredCharacters if after booted from an episode (Make playerStoredCharacters global), and have a shortened LaunchEpisode for Misc.loadEpisode

Commit:
Make world.getEpisodeList global everywhere

Commit:
Add GameAutostart things into Misc.loadEpisode

Commit:
Add GameAutostart.setSelectedEpisodePath

Commit:
Resort world-related global functions, bring GameAutostart functions from Misc.loadEpisode to LanuchEpisode itself

Commit:
Oops

Commit:
Forgot to remove this

Commit:
Add support for autostart.ini on boot, make error message when nothing is found a bit clearer

Commit:
Oops, again

Commit:
Add ANSI Windows path check to LaunchEpisode

Commit:
Remove COMBOOL on the if statements

Commit:
Massive commit. Not completely finished yet though

Commit:
Move LaunchEpisode to a new cpp/h file, add playerCount setting to GameAutostart, add support for world blocked characters, and so much more

Commit:
Epiodes now finally boot, but there's still a lot to do before I can consider this finished

Commit:
Make LunaDLL compilable again

Commit:
Remove level load stuff, will be moved to another branch

Commit:
CMD has a new argument ("--playSfxOnStartup"), which can be used to play a SFX when the episode has loaded on boot

Commit:
Fix potential issues with unused functions

Commit:
Fix up "tempLocation = WorldPath(A).Location" mess

Commit:
Fix for loop error

Commit:
Fix templates, add GM_SAVE_PERCENTAGE_PTR (For later)

Commit:
Extended way to check for a wld file

Commit:
Fix compile issues

Commit:
More code reworking mess. Also added EpisodeMain::FindSaves

Commit:
Oops

Commit:
Idk why world.getEpisodeList is nil... will have to check later

Commit:
Make episode setting values private again, also fix Misc.loadEpisode not loading the proper wld file when specifying a name

Commit:
Merge branch 'WohlSoft:master' into faster-episode-booting

Commit:
Remove unused LunaLua loadEpisode function

Commit:
Merge branch 'faster-episode-booting' of https://github.com/SpencerEverly/LunaLua into faster-episode-booting

Commit:
Merge branch 'WohlSoft:master' into faster-episode-booting

Commit:
Fix issue with episode booting wrong

Commit:
Move getEpisodeList to Misc, and fix no world existence error

Commit:
Extend Misc.loadEpisode to also have a save slot, number of players, and other player IDs of a higher player count specified

Commit:
Specify GM_MAX_CHARACTERS instead of just 5, and fix something with the first/second characters that may not apply them correctly

Commit:
Move the private variables from GameAutostart back to the top

Commit:
Remove unused canExecuteViaLua variables

Commit:
Fix some line spaces and comments

Commit:
External wld files on Misc.loadEpisode can now be specified, but just to warn you: Some episodes might crash.

Commit:
Fix possible episodeName issue on Misc.loadEpisode

Commit:
Forgot this

Commit:
Fix things for episodes if the GM_EP_LIST_COUNT is more than 100, also attempt to make external episode booting better but it's not there yet

Commit:
Actually fix more than 100 episode booting

Commit:
Attempt to fix everything to satisfy the devs (Part 1, after this commit I'll merge the latest LunaLua commit)

Commit:
Merge remote-tracking branch 'upstream/master' into faster-episode-booting

Commit:
Make it compilable again

Commit:
Merge remote-tracking branch 'upstream/master' into faster-episode-booting

Commit:
Put this pull request's new GlobalFuncs functions in the cpp file to below the file to prevent commit merge breaking changes

Commit:
See description for changes

- On the fast episode boot source codes & GlobalFunc.cpp fast episode boot source codes: Switch old GM, old native functions, and old SMBXInternal things to the new SMBX13 namespace environment
- Port "Misc.getEpisodeList" to a new LuaBind namespace ("Episode"), along with adding some new world-related functions, as well as renaming the functions to be shorter and easier to remember ("Misc.getEpisodeList()" is now "Episode.list()", and there's also "Episode.id()", "Episode.name()", "Episode.path()", and "Episode.filename()")
- Replace "WStr2Str(gAppPathWCHAR)" with gAppPathUTF8 in the checkIfWorldIsIn* functions (It looks better this way)
- Make episode loading more better, it hopefully can now withstand more than 100 episodes in the worlds folder (Though it's still capped at 100)

Commit:
Remove useless "For(A, From, To)" function, since it's now unused

Commit:
Merge remote-tracking branch 'upstream/master' into faster-episode-booting

ds-sloth commented:
Cleaning up old PRs in preparation to migrate to the new dev team
